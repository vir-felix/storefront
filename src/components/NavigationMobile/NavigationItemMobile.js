import React, { Component, Fragment } from "react";
import PropTypes from "prop-types";
import { inject } from "mobx-react";
import { Router } from "routes";
import ListItemIcon from "@material-ui/core/ListItemIcon";
import ListItemText from "@material-ui/core/ListItemText";
import Divider from "@material-ui/core/Divider";
import Collapse from "@material-ui/core/Collapse";
import MenuList from "@material-ui/core/MenuList";
import MenuItem from "@material-ui/core/MenuItem";
import ChevronDownIcon from "mdi-material-ui/ChevronDown";
import ChevronUpIcon from "mdi-material-ui/ChevronUp";
import { withStyles } from "@material-ui/core/styles";

const styles = (theme) => ({
  subNav: {
    marginBottom: theme.spacing.unit * 2
  },
  listItemTextInset: {
    "&:first-child": {
      paddingLeft: theme.spacing.unit * 3
    }
  }
});

@inject("routingStore")
@withStyles(styles, { name: "SkNavigationItemMobile" })
class NavigationItemMobile extends Component {
  static propTypes = {
    classes: PropTypes.object,
    isTopLevel: PropTypes.bool,
    navItem: PropTypes.object,
    onClick: PropTypes.func,
    routingStore: PropTypes.object
  };

  static defaultProps = {
    classes: {},
    navItem: {},
    onClick() {},
    routingStore: {}
  };

  state = { isSubNavOpen: false };

  get linkPath() {
    const { navItem, routingStore } = this.props;
    return routingStore.queryString !== ""
      ? `/tag/${navItem.slug}?${routingStore.queryString}`
      : `/tag/${navItem.slug}`;
  }

  get hasSubNavItems() {
    const { navItem: { subTags } } = this.props;
    return Array.isArray(subTags.edges) && subTags.edges.length > 0;
  }

  onClick = () => {
    const { navItem, isTopLevel } = this.props;

    if (isTopLevel && this.hasSubNavItems) {
      this.props.onClick(navItem);
    } else if (this.hasSubNavItems) {
      this.setState({ isSubNavOpen: !this.state.isSubNavOpen });
    } else {
      const path = this.linkPath;
      Router.pushRoute(path, { slug: navItem.slug });
    }
  };

  onClose = () => {
    this.setState({ isSubNavOpen: false });
  };

  renderSubNav() {
    const { classes, navItem: { subTags } } = this.props;
    return (
      <Collapse in={this.state.isSubNavOpen} timeout="auto" unmountOnExit>
        <MenuList component="div" disablePadding dense>
          {subTags.edges.map(({ node: navItemGroup }, index) => (
            <MenuItem className={classes.nested} key={index}>
              <ListItemText classes={{ inset: classes.listItemTextInset }} inset primary={navItemGroup.name} />
            </MenuItem>
          ))}
        </MenuList>
      </Collapse>
    );
  }

  render() {
    const { classes, navItem } = this.props;
    return (
      <Fragment>
        <MenuItem color="inherit" onClick={this.onClick}>
          <ListItemText classes={{ primary: classes.primary }} primary={navItem.name} />
          {this.hasSubNavItems && (
            <ListItemIcon className={classes.icon}>
              {this.state.isSubNavOpen ? <ChevronUpIcon /> : <ChevronDownIcon />}
            </ListItemIcon>
          )}
        </MenuItem>
        {this.hasSubNavItems && this.renderSubNav()}
        <Divider />
      </Fragment>
    );
  }
}

export default NavigationItemMobile;
