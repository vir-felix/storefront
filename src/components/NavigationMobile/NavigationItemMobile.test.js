import React from "react";
import renderer from "react-test-renderer";
import NavigationItemMobile from "./NavigationItemMobile";

const testMenuItem = {
  _id: "123",
  name: "Shop",
  slug: "shop"
};

const testMenuItemWithRelatedTags = {
  _id: "123",
  name: "Shop",
  slug: "shop",
  relatedTags: [
    { _id: "111", name: "Clothes", slug: "clothes" },
    { _id: "222", name: "Shoes", slug: "shoes" },
    { _id: "333", name: "Accessories", slug: "accessories" }
  ]
};

test("basic snapshot", () => {
  const component = renderer.create(<NavigationItemMobile navItmes={testMenuItem} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("snapshot with related items", () => {
  const component = renderer.create(<NavigationItemMobile navItmes={testMenuItemWithRelatedTags} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
