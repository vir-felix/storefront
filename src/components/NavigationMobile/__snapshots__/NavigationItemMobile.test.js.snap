// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with 1 level of tags 1`] = `null`;

exports[`renders with 2 levels of tags 1`] = `
<li
  className="MuiButtonBase-root-15 MuiListItem-root-5 MuiListItem-default-8 MuiListItem-gutters-12 MuiListItem-button-13 MuiMenuItem-root-3"
  color="inherit"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  role="menuitem"
  tabIndex={-1}
>
  <div
    className="MuiListItemText-root-18"
  >
    <span
      className="MuiTypography-root-24 MuiTypography-subheading-31 MuiListItemText-primary-21"
    >
      Tag Name
    </span>
  </div>
  <svg
    aria-hidden="true"
    className="MuiSvgIcon-root-49 MuiListItemIcon-root-48"
    color={undefined}
    focusable="false"
    viewBox="0 0 24 24"
  >
    <g>
      <path
        d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"
      />
    </g>
  </svg>
  <span
    className="MuiTouchRipple-root-56"
  />
</li>
`;

exports[`renders with 3 levels of tags 1`] = `
<li
  className="MuiButtonBase-root-15 MuiListItem-root-5 MuiListItem-default-8 MuiListItem-gutters-12 MuiListItem-button-13 MuiMenuItem-root-3"
  color="inherit"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  role="menuitem"
  tabIndex={-1}
>
  <div
    className="MuiListItemText-root-18"
  >
    <span
      className="MuiTypography-root-24 MuiTypography-subheading-31 MuiListItemText-primary-21"
    >
      Tag Name
    </span>
  </div>
  <svg
    aria-hidden="true"
    className="MuiSvgIcon-root-49 MuiListItemIcon-root-48"
    color={undefined}
    focusable="false"
    viewBox="0 0 24 24"
  >
    <g>
      <path
        d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"
      />
    </g>
  </svg>
  <span
    className="MuiTouchRipple-root-56"
  />
</li>
`;
